### ISSUE TITLE
saw htmlwidget on the todo

### ISSUE THREAD

**Issue by timelyportfolio** [cmpolis/datacomb#1]
I saw `htmlwidget` on the todo and I'd love to help.  I will spend some time today getting acquainted with this fine piece of work.

**Comment by jbkunst** [comment]
:+1:

**Comment by cmpolis** [comment]
Terrific - thanks for your interest and help! The demo code is not available - this project is a complete rewrite, which will hopefully be usable soon. I'm excited about having an R interface and being able to easily plug in data frames.

I haven't come up with an api or thought about the implementation, but I'd imagine it being similar to the [DataTable htmlwidget](http://www.htmlwidgets.org/showcase_datatables.html). I'm a bit of an R novice, so any help is greatly appreciated! :raised_hands:

**Comment by abresler** [comment]
This would be fantastic - happy to try to help as well.  Try to taking a look at the very nice port of [PivottableJS](https://github.com/nicolaskruchten/pivottable) which made its way into a great htmlwidget you can see [here](https://github.com/smartinsightsfromdata/rpivotTable)

**Comment by timelyportfolio** [comment]
I see that you are making solid progress.  When I try to run on my side, I get this:

![image](https://cloud.githubusercontent.com/assets/837910/9976130/72704cf0-5e9f-11e5-8e4a-343a0ca6b455.png)

The constructor on my side would be `new Datacomb({...})` but then I get other errors.  Am I missing something obvious before I dive deep into the code?

**Comment by cmpolis** [comment]
Thanks, things are coming along pretty nicely!

It looks like the first screenshot was coming from things not being built, does `$ npm run build` complete without error *(this step was previously not in the Readme, sorry!)*? I will document this better, but for now the best sample of how to init an interface is in `/demo/datacomb/demo.js`

**Comment by cmpolis** [comment]
Also, I have some `htmlwidget/R` code from a previous version of datacomb that I will try to carry over to this iteration, this week!

**Comment by timelyportfolio** [comment]
Something on my end, so now the demo is working, and it is beautiful.  I'd love to help implement the `htmlwidget`.  I guess following along on Github will be best.

**Comment by abresler** [comment]
Oh this would be so so amazing!!

----------------------
Alex Bresler
abresler@asbcllc.com
​www.asbcllc.com​
917-455-0239​ (cell)​

**Comment by cmpolis** [comment]
I put in a quick, first pass for this! I have a working knowledge of R, but am somewhat clueless on defining the proper API, docs, etc... so feel free to put in updates! Currently, the API is fairly basic; it just consumes a dataframe and column and label params. Also, let me know if you are able to install and get this up and running through R.

```R
devtools::install_github('cmpolis/datacomb')
library(datacomb)
Datacomb(iris, columns=c('Sepal.Length','Sepal.Width', 'Petal.Length', 'Petal.Width', 'Species'))
```

**Comment by timelyportfolio** [comment]
I'll start playing with it now. This looks like a great start.

**Comment by timelyportfolio** [comment]
@cmpolis I toyed with it a bit at https://github.com/timelyportfolio/datacomb/tree/R.

```
# devtools::install_github('timelyportfolio/datacomb@R', subdir = 'pkg')

library(datacomb)
?Datacomb
Datacomb(mtcars)

data(diamonds, package = "ggplot2")
Datacomb(diamonds)
```

**Comment by abresler** [comment]
Guys had to just chime in and say this is every bit as amazing I hoped; holy CRAP

**Comment by cmpolis** [comment]
@timelyportfolio :+1: terrific - thanks for the updates!!! Can I merge in/do you want to put in a pr?

**Comment by timelyportfolio** [comment]
@cmpolis, easy when you've done such an amazing job.  I hesitated to submit pr before making sure that a blended repo wtih JS/node + R + maybe python + who knows what other implementations would be a good strategy, or if each implementation should have its own repo to avoid confusion.  Would a `datacomb` organization be an option?

**Comment by cmpolis** [comment]
I would shy away against creating an organization, just because it will add some complexity in terms of managing repos, users, what is the process for bringing a repo into the org, etc... the main non-web interface that I am interested in is R; I think that's fine to leave in here. The list other potential interfaces/package management systems is somewhat endless, so I'd like to not get too involved and focus on the actual interface :sunglasses: (and `R` functionality b/c I will be using this myself). If others want to make repos that plug into datacomb or put in a pr for `[bower, component,...].json`, that's great!

**Comment by cmpolis** [comment]
that being said, one of the great things about your pr is having everything up a level, in `/pkg`! I will likely reorganize this project to have a more san dir structure:
```
/src/[js,css,hbs,etc...]/
/dist
/lib
/test
...
```

**Comment by timelyportfolio** [comment]
Definitely helps to understand this, and I appreciate the insight.  `/pkg` requires the `subdir` argument, but I think the advantages will be worth the extra argument, especially if you are thinking of a san dir structure.  Do you want me to wait to see the restructure before a pull?

**Comment by cmpolis** [comment]
+1 don't mind `subdir` arg; I think most people will be copy/pasting or looking at some sample code anyways to install. Feel free to put in your pr - I'll cleanup and do my refactoring later.

Few things re the `DESCRIPTION` file:
- can we remove the `R` dependency or lower the version number? I had a friend try to install this today and was seeing issues because of a fairly recent version being required.
- what else should we do with the `depends`, `imports`... properties?
- please add yourself on there!

**Comment by timelyportfolio** [comment]
This doesn't seem to be working anymore http://timelyportfolio.github.io/buildingwidgets/week38/example01.html.  Scrolling stops at 1000 records.

**Comment by cmpolis** [comment]
Yes, there seems to be an issue w/ row reuse. If you bump [this parameter](https://github.com/cmpolis/datacomb/blob/master/src/js/datacomb.js#L146), I think it will be fixed.

**Comment by cmpolis** [comment]
@timelyportfolio I think this might fix (or fix part of) your issue: https://github.com/cmpolis/datacomb/commit/e730b148e828849912d68a875801b1a40bf2870f ... still working out some kinks in this thing

**Comment by timelyportfolio** [comment]
Yes, thanks that fixes the `div` recycle, but I now seem to have a new problem.  I'll try to describe it in words and illustrate with a GIF.  Focus on hover works great until we scroll past the initial available `div` pool.  Then focus on hover moves to the bottom of the screen, and the screen becomes blank.

![datacomb_focus_problem](https://cloud.githubusercontent.com/assets/837910/10100534/172eebae-6358-11e5-9085-e8e79e566ed2.gif)


I'm going to try to debug, but afraid my skills won't be sufficient for the task.

**Comment by timelyportfolio** [comment]
Maybe this is a clue as style goes from > 2000000px to 0px.  I'll keep digging.

![datacomb_focus_problem_elements](https://cloud.githubusercontent.com/assets/837910/10101324/a4d57176-635d-11e5-8e13-407e1f9ed844.gif)

**Comment by timelyportfolio** [comment]
I think I have traced the problem to this [line](https://github.com/cmpolis/smart-table-scroll/blob/master/src/smart-table-scroll.js#L102) in smart-table-scroll because this [line](https://github.com/cmpolis/smart-table-scroll/blob/master/src/smart-table-scroll.js#L178) sets top to `0px`++ and then I think expects [line](https://github.com/cmpolis/smart-table-scroll/blob/master/src/smart-table-scroll.js#L182) to properly set `top`.

**Comment by cmpolis** [comment]
Thanks for the detailed report; looking into it!
